cmd/clusterctl/client/alpha/rollout_viewer.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpha
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	logf "sigs.k8s.io/cluster-api/cmd/clusterctl/log"
+	"sigs.k8s.io/cluster-api/internal/controllers/machinedeployment/mdutil"
+)
+
+// ObjectViewer will issue a view on the specified cluster-api resource.
+func (r *rollout) ObjectViewer(proxy cluster.Proxy, ref corev1.ObjectReference, revision int64) error {
+	switch ref.Kind {
+	case MachineDeployment:
+		deployment, err := getMachineDeployment(proxy, ref.Name, ref.Namespace)
+		if err != nil || deployment == nil {
+			return errors.Wrapf(err, "failed to get %v/%v", ref.Kind, ref.Name)
+		}
+		if err := viewMachineDeployment(proxy, deployment, revision); err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("invalid resource type %q, valid values are %v", ref.Kind, validResourceTypes)
+	}
+	return nil
+}
+
+func viewMachineDeployment(proxy cluster.Proxy, d *clusterv1.MachineDeployment, revision int64) error {
+	log := logf.Log
+	msList, err := getMachineSetsForDeployment(proxy, d)
+	if err != nil {
+		return err
+	}
+
+	if revision < 0 {
+		return errors.Errorf("revision number cannot be negative: %v", revision)
+	}
+
+	// Print details of a specific revision
+	if revision > 0 {
+		ms, err := findMachineDeploymentRevision(revision, msList)
+		if err != nil {
+			return errors.Errorf("unable to find the spcified revision")
+		}
+		output, err := yaml.Marshal(ms.Spec.Template)
+		if err != nil {
+			return err
+		}
+		fmt.Fprint(os.Stdout, string(output))
+		return nil
+	}
+
+	// Print an overview of all revisions
+	// Create a revisionToChangeCause map
+	historyInfo := make(map[int64]string)
+	for _, ms := range msList {
+		v, err := mdutil.Revision(ms)
```

Comment:
I just stumbled into machinedeployment.clusters.x-k8s.io/revision-history and I'm wondering if we should keep this as well into account while building the revision history

Reply:
Thank you for your comments. 
Maybe, it's a good time to discuss the deprecation of `revision-history`. I personally can't find any use cases for this annotation but we need to confirm that, and I think it should be handled by another issue. 

FYI, I found that Kubernetes also has [this annotation](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L50), but it's not used in the actual logic.

Reply:
In CAPI it is used in [https://github.com/kubernetes-sigs/cluster-api/blob/29618402670e19c77833edbede6d9e[…]7b4d1036a/internal/controllers/machinedeployment/mdutil/util.go](https://github.com/kubernetes-sigs/cluster-api/blob/29618402670e19c77833edbede6d9e87b4d1036a/internal/controllers/machinedeployment/mdutil/util.go#L222-L234)

TL;DR each machineset keeps track of the revision it currently is, and if “reused”, it keeps also track of the revision it was…

Reply:
I would prefer not showing the revision numbers that are listed under revision history annotation, at least as separate entries in the output. 
Since the idea behind the history command is to list the revisions the user can rollback to, it would be best to not show the revision numbers in the revision history annotation as it is not possible to rollback using those revision numbers. 

If we want really want to show those numbers in the output for the sake of having a full history then how about adding a third column called "matching revisions"/"duplicate reivisions", or some better name where we list these revision numbers against target revision numbers?

Thoughts?

Reply:
Why can't we rollback to revision number in revision history?

Reply:
Following up from the discussion in the office hours.

I think it's very user unfriendly if we do not support revision-history. Today Cluster API only tracks MachineSets that are actually different via separate MachineSet objects. As soon a MachineSet is the same as another one in the history revision-history is used automatically.

If we don't support revision-history from a user point-of-view there will be arbritrary holes in the history (e.g. if 1, 4, 7 are the same the history is: 2, 3, 5, 6, 8). I think this is very hard to understand for users.

I would suggest:
* in the history command: list all revisions including the one from the revision-history annotation
* for the rollback command: make it possible to also rollback to revisions from the revision history (should be as easy as also looking at the revision-history annotation instead of only at the revision annotation to find the MachineSet for a specific revision)

Reply:
Thank you for following up and My apologies for the delayed reply.

I changed PR following your suggestion. 

I'll deal with your second suggestion in a different PR as it should be an enhancement of `undo`.

> should be as easy as also looking at the revision-history annotation instead of only at the revision annotation to find the MachineSet for a specific revision

This makes sense.

---

BTW, I made output of the history command like this. If I am a user, I want to group revisions associated to the same MachineSet with the same change cause.

```bash
REVISIONS       CHANGE-CAUSE
        2             <none>  
    4,6,8,10      add label
   1,3,5,7,9,11   update to the latest k8s version
```

The above is more useful than this when you are looking for a revision that you want to rollback.

```bash
REVISION        CHANGE-CAUSE
        1             <none>  
        2             <none>  
        3             <none>  
        4             <none>  
        5             <none>  
        6             <none>  
        7             <none>  
        8             <none>  
        9             <none>  
       10            add label  
       11            update to the latest k8s version
```

As an alternative, we can extend the change cause to record its history but I think that is too much.